src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Parser.java

```diff
@@ -0,0 +1,496 @@
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.javadoc.internal.doclets.toolkit.taglets.snippet;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+/*
+ * Semantics of a EOL comment; plus
+ * 1. This parser treats input as plain text. This may result in markup being
+ * produced from unexpected places; for example, when parsing Java text blocks:
+ *
+ *     String text =
+ *         """
+ *             // @start x
+ *         """;
+ *
+ * false positives are possible, but false negatives are not.
+ * To remediate that, perhaps a no-op trailing // @start x @end x might be added.
+ *
+ * 2. To allow some preexisting constructs, unknown actions in a leading position are skipped;
+ * for example, "// @formatter:on" marker in IntelliJ IDEA is ignored.
+ *
+ * 3. This match's value can be confused for a trailing markup.
+ *
+ *     String x; // comment // another comment // @formatter:on // @highlight match="// @"
+ *
+ * Do we need escapes?
+ *
+ * 4. Rules for EOL are very different among formats: compare Java's // with properties' #/!
+ *
+ * 5. A convenience `end` ends all the things started so far.
+ */
+/**
+ * A parser of snippet content.
+ *
+ * <p><b>This is NOT part of any supported API.
+ * If you write code that depends on this, you do so at your own risk.
+ * This code and its internal interfaces are subject to change or
+ * deletion without notice.</b>
+ */
+public final class Parser {
+
+    // next-line tag behaves as if it were specified on the next line
+
+    private String eolMarker;
+    private Matcher markedUpLine;
+    private final MarkupParser markupParser = new MarkupParser();
+
+    // Incomplete actions waiting for their complementary @end
+    private final Regions regions = new Regions();
+    // List of tags; consumed from top to bottom
```

Comment:
"top to bottom" is somewhere between ambiguous and confusing.  Is it file-centric or queue-as-stack -centric?